lang/ast/structs.go

```diff
@@ -8941,6 +8942,7 @@ func (obj *ExprTopLevel) Value() (types.Value, error) {
 type ExprSingleton struct {
 	Definition interfaces.Expr
 
+	lock           sync.Mutex // protect singletonGraph and singletonExpr
```

Comment:
You'll want a *sync.Mutex and for it to be initialized somewhere... Probably in Init() if it's only ever init-ed once.

Reply:
Hmm, I get `panic: runtime error: invalid memory address or nil pointer dereference`, is it possible that Init() is never called? Or maybe am I initializing the mutex wrong?

Reply:
You init it with:

mutex = &sync.Mutex{}

Reply:
that's what I wrote in Init() (or rather that's what GitHub Copilot wrote in Init()), so Init() probably isn't called. which part of the code is responsibile for calling Init()?

Reply:
> which part of the code is responsibile for calling Init()?

This answer is not trivial, which is why I'm happy to do that figuring out and plumbing on your behalf if you'd like, since you're better off working on more important things that a dumb plumber like me can't figure out! :sweat: 

(The short answer is the top-level Init recursively calls everything, but there are a few places where we should probably be running Init that we might not be yet. And it's more complicated because of the Init data structure we pass through everywhere...)

Reply:
> I'm happy to do that figuring out and plumbing on your behalf if you'd like

yes please!